### PR TITLE
 [HW-3] Solve 4 tasks

### PR DESCRIPTION
**Pull request by lizamuromskaya**

### CODE REVIEW

File: HW03/Lizaveta_Muromskaya/liza_muromskaya_hw_03_runner.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+require_relative 'lizaveta_muromskaya_hw_03_t_01'
+require_relative 'lizaveta_muromskaya_hw_03_t_02'
+require_relative 'lizaveta_muromskaya_hw_03_t_03'
+require_relative 'lizaveta_muromskaya_hw_03_t_04'
+
+log1 = '10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290'
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/liza_muromskaya_hw_03_runner.rb:9]:
```suggestion
log1 = <<~LOGS
  10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
  10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
  2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
  10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
LOGS
```

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_01.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  arr = log.split("\n")
+  count = arr.length - 1
+  arr.each_with_index do |i, c|
+    if i['error']
+      p i
+      break
+    elsif c == count && !i['error']
+      p "\n"
+      break
+    end
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_01.rb:11]: Find and check what the "include" method is used for

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+def task_2(log)
+  @arr = []
+  str = ''
+  data = log.scan(%r{[0-9]+/[A-z]+/[0-9]+:+[0-9]+:+[0-9]+:+[0-9]+\s+\+[0-9]*}) # data
+  from = log.scan(/^[0-9]+\.+[0-9]+\.+[0-9]+\.+[0-9]*/) # from
+  to = log.scan(%r{/[a-z]+/[0-9]+/[a-z]+}) # to
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb:6]: Could you move regex to constants please?

Comment (B0RlS) [reply HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb:6]: and remove comments please

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+def task_2(log)
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb:1]: Why don't you go through the array for each line of the log and apply regular expressions for each line.
This implementation will be easier

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  data = log.scan(%r{[0-9]+/[A-z]+/[0-9]+:+[0-9]+:+[0-9]+:+[0-9]+\s+\+[0-9]*}) # data
+  from = log.scan(/^[0-9]+\.+[0-9]+\.+[0-9]+\.+[0-9]*/) # from
+  to = log.scan(%r{/[a-z]+/[0-9]+/[a-z]+}) # to
+  to.each { |i| i.upcase! }
+  (0..data.size - 1).each do |i|
+    str = "DATA: #{data[i]} FROM: #{from[i]} TO: #{to[i]}"
+    @arr << str
+  end
+  if @arr.size > 0
+    @arr.each { |i| p i }
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb:13]:
```suggestion
    @arr.each { |i| p i }
```
`puts arr` is equal to your code

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+def task_3(log)
+  @arr = []
+  @result = []
+  # Calling core with action:
+  a = log.split("\n")
+  a.each_with_index do |string, _index|
+    if string['Calling core with action:']
+      time = string.scan(/[0-9]+:[0-9]+:[0-9]+\.[0-9]/) # time
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:8]: Could you move regex to constants please?

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb (hunk 4 above)
Comment (B0RlS) [reply HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_02.rb:1]: In your 3rd assignment, you did it

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+def task_3(log)
+  @arr = []
+  @result = []
+  # Calling core with action:
+  a = log.split("\n")
+  a.each_with_index do |string, _index|
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:6]: And why do you need an iterator with an index? I think you can use a regular iterator

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    if string['Calling core with action:']
+      time = string.scan(/[0-9]+:[0-9]+:[0-9]+\.[0-9]/) # time
+    end
+    next if time.nil?
+
+    time.each do |i|
+      d = i.split(':')
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:13]: Could you not name the variables with a single letter please?

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      sec = d[0].to_f * 3600 + d[1].to_f * 60 + d[2].to_f
+      @arr << sec
+    end
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:16]:
```suggestion
    time.each do |i|
      d = i.split(':')
      sec = d[0].to_f * 3600 + d[1].to_f * 60 + d[2].to_f
      @arr << sec
    end
```
Could you move this code into another method and call it please?

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    (0..@arr.size - 2).each do |i|
+      res = @arr[i + 1] - @arr[i]
+      @result << res.round(1).to_s if res > 0
+    end
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:20]: Could you explain to me what is going on here and why you are doing `(0..@arr.size - 2)`?

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_04.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+def task_4(str)
+  hh = {}
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_04.rb:2]: Could you rename the variable and give it a more descriptive name please?

**Comment by B0RlS** [comment]
Could you run `rubocop -a` in your HW folder please?
Screenshot 2021-04-28 at 10 32 30

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_01.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+def task1(log)
+  lines = log.split("\n")
+  lines.each { |line| return p line if line.include?('error') }
+  p ''
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_01.rb:6]:
```suggestion
  lines.each { |line| return line if line.include?('error') }
  ''
```

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_04.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+def task4(data)
+  character_counter = {}
+  character_counter[:letters] = data.count('A-Za-z')
+  character_counter[:digits] = data.count('0-9')
+  p character_counter
```
Comment (B0RlS) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_04.rb:7]:
```suggestion
  character_counter
```

**Review by B0RlS — APPROVED** [review]
LGTM

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    sec = time_buf[0].to_f * 3600 + time_buf[1].to_f * 60 + time_buf[2].to_f
+    @seconds << sec
+  end
+end
+
+def time_difference
+  @seconds.each_cons(2) do |start_event, end_event|
+    res = end_event - start_event
+    @result << res.round(1).to_s if res.positive?
+  end
```
Comment (vladhilko) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:16]: Methods shouldn't create any side effects ( like updating global variables etc ). You have 2 options here:
- wrap everything into a class  ( inside the class it's ok )
- just send argument and return value 
```suggestion
def time_difference(seconds)
  seconds.each_cons(2).map do |start_event, end_event|
    res = end_event - start_event
    res.positive? ? res.round(1).to_s : nil
  end.compact
```

File: HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb [hunk 15]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+TIME = /[0-9]+:[0-9]+:[0-9]+\.[0-9]/.freeze
+def time_in_sec(time, seconds)
+  time.each do |time_numbers|
+    time_buf = time_numbers.split(':')
+    sec = time_buf[0].to_f * 3600 + time_buf[1].to_f * 60 + time_buf[2].to_f
+    seconds << sec
+  end
+end
```
Comment (vladhilko) [inline HW03/Lizaveta_Muromskaya/lizaveta_muromskaya_hw_03_t_03.rb:10]: This method also shouldn't modify `seconds` 😉 
```suggestion
def time_in_sec(time)
  time.map do |time_numbers|
    time_buf = time_numbers.split(':')
    time_buf[0].to_f * 3600 + time_buf[1].to_f * 60 + time_buf[2].to_f
  end
end

seconds = time_in_sec(time)
```

**Review by vladhilko — APPROVED** [review]

**Pull request merged**